mlx/backend/metal/indexing.cpp

```diff
@@ -641,4 +643,154 @@ void ScatterAxis::eval_gpu(const std::vector<array>& inputs, array& out) {
+
+  scan_gpu(
+      mask_flat,
+      scatter_offsets,
+      Scan::Sum,
+      /*axis=*/0,
+      /*reverse=*/false,
+      /*inclusive=*/false,
+      s,
+      /*allow_in_buffer_donation=*/false);
+
+  // Kernel selection/build
+  const bool is_batched_run = batched && (batch > 1);
+  const bool src_has_batches = is_batched_run && src_batched;
+  static constexpr std::string_view kBaseName = "masked_assign";
+  const std::string dtype_tag = type_to_name(out.dtype());
+  const std::string value_type = get_type_string(out.dtype());
+  const std::string kernel_name = fmt::format(
+      "{}_{}_b{}_src{}",
+      kBaseName,
+      dtype_tag,
+      is_batched_run ? 1 : 0,
+      src_has_batches ? 1 : 0);
+
+  auto lib = d.get_library(kernel_name, [&]() {
+    std::string source = metal::utils();
+    source += metal::masked_scatter();
+    source += fmt::format(
+        std::string(masked_assign_kernel),
+        kernel_name,
+        value_type,
+        is_batched_run ? "true" : "false",
+        src_has_batches ? "true" : "false");
+    return source;
+  });
+  auto kernel = d.get_kernel(kernel_name, lib);
+
+  // Binding
+  int bind_idx = 0;
+  auto& compute_encoder = d.get_command_encoder(s.index);
+  compute_encoder.set_compute_pipeline_state(kernel);
+  compute_encoder.set_input_array(mask_flat, bind_idx++);
+  compute_encoder.set_input_array(scatter_offsets, bind_idx++);
+  compute_encoder.set_input_array(src, bind_idx++);
+  compute_encoder.set_output_array(out, bind_idx++);
+
+  auto bind_shape_info =
+      [&](const array& arr, int shape_index, int stride_index, int ndim_index) {
+        const int ndim = static_cast<int>(arr.ndim());
+        if (ndim == 0) {
+          const int shape0 = 0;
+          const int64_t stride0 = 0;
+          compute_encoder.set_bytes(shape0, shape_index);
+          compute_encoder.set_bytes(stride0, stride_index);
+        } else {
+          compute_encoder.set_vector_bytes(arr.shape(), shape_index);
+          compute_encoder.set_vector_bytes(arr.strides(), stride_index);
+        }
+        compute_encoder.set_bytes(ndim, ndim_index);
+      };
+
+  bind_shape_info(mask, bind_idx, bind_idx + 1, bind_idx + 2);
+  bind_idx += 3;
+  bind_shape_info(src, bind_idx, bind_idx + 1, bind_idx + 2);
+  bind_idx += 3;
+  bind_shape_info(out, bind_idx, bind_idx + 1, bind_idx + 2);
+  bind_idx += 3;
+
+  auto set_u32 = [&](uint32_t v, int idx) {
+    compute_encoder.set_bytes(v, idx);
+  };
+  auto as_u32 = [](size_t v) -> uint32_t {
+    if (v > std::numeric_limits<uint32_t>::max())
+      throw std::overflow_error(
+          "[MaskedScatter::eval_gpu] uint64 size not implemented.");
+    return static_cast<uint32_t>(v);
+  };
+  set_u32(as_u32(src_size), bind_idx++);
+  set_u32(as_u32(total), bind_idx++);
+  set_u32(is_batched_run ? as_u32(inner) : 0u, bind_idx++);
+  set_u32(src_has_batches ? as_u32(src_block) : 0u, bind_idx++);
+  set_u32(is_batched_run ? as_u32(src_capacity) : 0u, bind_idx++);
```

Comment:
Nit: I think the `set_u32` is unnecessary. Cleaner to just call `compute_encoder.set_bytes(...) ` directly.